src/client/renderer/r_bsp_light.c

```diff
@@ -322,3 +325,292 @@ void R_DrawBspLights(void) {
+
+			g_hash_table_insert(stain_table, surf->stainmap, lightmap);
+		}
+
+		// copy over the old stain buffer
+		const size_t stride = surf->lightmap->width * 3;
+		const size_t stain_width = surf->st_extents[0] * 3;
+		const size_t lightmap_offset = (surf->lightmap_t * surf->lightmap->width + surf->lightmap_s) * 3;
+
+		byte *light = lightmap + lightmap_offset;
+		byte *stain = surf->stainmap_buffer;
+
+		for (int16_t t = 0; t < surf->st_extents[1]; t++) {
+			
+			memcpy(stain, light, stain_width);
+
+			stain += stain_width;
+			light += stride;
+		}
+
+		surf->stainmap_dirty = false;
+	}
+
+	// free all the scratch
+	g_hash_table_destroy(stain_table);
+}
+
+/**
+ * @brief
+ */
+static void R_StainNode(const vec3_t org, const vec4_t color, const vec_t size, const r_bsp_node_t *seed_node, uint32_t *num) {
+
+	static GArray *node_list;
+	const vec_t stain_alpha = color[3] * r_stain_map->value;
+	const vec_t dst_stain_alpha = 1.0 - stain_alpha;
+
+	if (!node_list) {
+		node_list = g_array_new(false, false, sizeof(r_bsp_node_t *));
+	} else {
+		node_list->len = 0;
+	}
+
+	g_array_append_val(node_list, seed_node);
+
+	while (node_list->len) {
+		const r_bsp_node_t *node = g_array_index(node_list, const r_bsp_node_t *, --node_list->len);
+
+		if (node->contents != CONTENTS_NODE) {
+			continue;
+		}
+
+		vec_t dist;
+		if (node->plane->type < SIDE_BOTH) {
+			dist = org[node->plane->type] - node->plane->dist;
+		} else {
+			dist = DotProduct(org, node->plane->normal) - node->plane->dist;
+		}
+
+		if (dist > size) {
+			g_array_append_val(node_list, node->children[0]);
+			continue;
+		} else if (dist < -size) {
+			g_array_append_val(node_list, node->children[1]);
+			continue;
+		}
+
+		r_bsp_surface_t *surf = r_model_state.world->bsp->surfaces + node->first_surface;
+
+		for (uint32_t c = node->num_surfaces; c; c--, surf++) {
+
+			if (!surf->lightmap || !surf->stainmap) {
+				continue;
+			}
+
+			const r_bsp_texinfo_t	*tex = surf->texinfo;
+
+			if ((tex->flags & (SURF_SKY | SURF_BLEND_33 | SURF_BLEND_66 | SURF_WARP))) {
+				continue;
+			}
+
+			vec_t fdist = DotProduct(org, surf->plane->normal) - surf->plane->dist;
+
+			if (surf->flags & R_SURF_PLANE_BACK) {
+				fdist *= -1;
+			}
+
+			const vec_t frad = size - fabs(fdist);
+			vec_t fminlight = 0.0;
+
+			if (frad < fminlight) {
+				continue;
+			}
+
+			fminlight = frad - fminlight;
+
+			vec3_t impact;
+			VectorMA(org, -fdist, surf->plane->normal, impact);
+
+			const vec2_t local = {
+				DotProduct (impact, tex->vecs[0]) + tex->vecs[0][3] - surf->st_mins[0],
+				DotProduct (impact, tex->vecs[1]) + tex->vecs[1][3] - surf->st_mins[1]
+			};
+
+			const r_pixel_t smax = surf->st_extents[0];
+			const r_pixel_t tmax = surf->st_extents[1];
+
+			vec_t ftacc = 0.0;
+
+			byte *pfBL = surf->stainmap_buffer;
+			_Bool changes = false;
+
+			for (uint16_t t = 0; t < tmax; t++, ftacc += 16.0) {
```

Comment:
I think you want to increment by lightmap_scale here, instead of 16.0. Might this be part of the problem with scaling stains correctly?